### DISCUSSION TITLE
How to deal with nested levels of `Unis`

### DISCUSSION THREAD

**Discussion by Moicen** [quarkusio/quarkus#22690] [Community]
I need to implement a `@ServerRequestFilter`, and it calls several levels of async calls, which generates nested  `Unis`, and here is the screenshot of the code:

![C94C3529-972F-4262-A54D-0AC505B5B0F0](https://user-images.githubusercontent.com/3773928/148394629-8f32215d-e298-4645-97df-8ff0d5e905d0.png)

However this makes a problem that makes the return type has nested level of `Unis` :

![6684DDB2-5626-4A70-BA6B-39FB1E221783](https://user-images.githubusercontent.com/3773928/148394663-d2235076-a31c-4cbf-a43e-bdbd064d73df.png)

And this breaks the convention of `@ServerRequestFilter`:

![D53D46A1-A28B-464A-B452-B0FB781BF0AA](https://user-images.githubusercontent.com/3773928/148394685-6de99329-dd45-474b-b8b5-d0f09f83de09.png)

Is there any solution to my problem?

**Comment by Moicen** [comment]
If I add an `await()` to reduce one level of `Uni` for nested `Uni<Uni<...>>`:

![B9672FDD-741A-415F-9EBD-5762F90BC72E](https://user-images.githubusercontent.com/3773928/148396983-e28b64b8-2e4b-4b8e-9bb2-e37fb7a4481b.png)

Then it generates following problem:

![76D62DBF-1007-4C36-8F37-3EAC181AEFC2](https://user-images.githubusercontent.com/3773928/148397008-a35bbcbc-fd85-4743-b544-4e5e96f8a05f.png)

So it’s not a possible solution.

**Comment by Moicen** [comment]
Here are the dependencies I’m using:

```xml
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-rest-client-reactive</artifactId>
</dependency>
<dependency>
    <groupId>io.quarkus</groupId>
    <artifactId>quarkus-resteasy-reactive</artifactId>
</dependency>
```

In addition I use `ReactiveRedisClient`:

![DB2F8702-72FE-4F57-9C2D-930F7B0E3D94](https://user-images.githubusercontent.com/3773928/148402141-a094f2ba-89ee-491b-b54f-0461bdb8cc39.png)

**Comment by benkard** [comment]
You can always do `uni.flatMap(Function.identity())` to go from `Uni<Uni<T>> uni` to `Uni<T>`.

In your particular case, it may be sufficient to replace one of the `.transform` calls with `.transformToUni`.

**Reply by Moicen** [reply]
I replaced `transform()` with `transformToUni()` and it works:

![50F73C1E-595F-4B55-91A3-08E601D62A76](https://user-images.githubusercontent.com/3773928/148491419-717afe15-d550-41b2-a145-be19b40a7c52.png)

Thanks for help!

**Comment by Moicen** [comment]
@benkard Now I have further problem:

![6B3A37E2-EB7A-45C5-8B67-9C436CDC8731](https://user-images.githubusercontent.com/3773928/148541892-48489305-f74c-4e99-9414-113ffd797832.png)

From above screenshot, you can see that by using `transformToUni()`, the above filter logic can work. But I have to comment out the `tokenCacheeService.setRequestSudoerToken(token)` method in the nested `uc.verifyJwtToken()` block.

If I enable it by using `transform()` method, and do a `return null` inside the block(here `subscribe().with()` together with a `sleep()` is not working because the `subscribe()` method doesn’t executed at all even after `sleep()` a while):

![9E94BC36-6092-4D6A-ABE8-302279E1E82C](https://user-images.githubusercontent.com/3773928/148541917-b3ec1ff1-0682-46ad-875b-2c631112bd0a.png)

 then it will produce the following error:

![97B96662-83D1-4669-9F36-873661D16EA9](https://user-images.githubusercontent.com/3773928/148541936-90e8f84b-884d-49b2-bb7e-0b979365f2c0.png)

This is because now there is one extra level of nested `onItem().invoke()` which makes the return type changed. So I use `transformToUni()` in this level but it generates following error:

![1A7A1730-A1A5-499E-A2C5-38FCB3A8320C](https://user-images.githubusercontent.com/3773928/148541956-ed06f597-bfd3-4587-a538-21b7ece5815c.png)

I have no idea how to further fixing it.

**Comment by Moicen** [comment]
I have separated out the root cause:

![BDC04B22-C578-4F46-AD5F-1739B66338C2](https://user-images.githubusercontent.com/3773928/148632715-0eeb4ac6-11ae-4072-a214-a8ec2fc68715.png)

From above screenshot you can see I have aligned two nested levels of `Uni<Uni<...>>`, and now the error becomes:

![588FA193-0ACC-4FAB-8CD4-3C9D1C0664DC](https://user-images.githubusercontent.com/3773928/148632720-8c765daa-da62-4aee-a3e1-5c1686e27f45.png)

So the `transformToUni()` seems not working for this two nested levels of `Unis`, is there any solution to this?

**Reply by benkard** [reply]
I don’t know what all the types are here, but I suspect that inner `.transform` again needs to be `.transformToUni`. 

Aside: Personally, I tend to split any such `Uni` pipeline into smaller methods and compose those with the `.transform` etc. combinators.  That way it is much easier to locate problems like this.  Method references are super useful here – my own code tends to look more like `return doThis.flatMap(this::doThat).flatMap(this::doAnotherThing);` and so on.

**Comment by Moicen** [comment]
@benkard  I simplified the code did some further tests, and here’s the result:

![DC654B18-4C15-42FF-8055-F9E11CD77014](https://user-images.githubusercontent.com/3773928/148647472-7bc724dd-6cb4-4826-b7fa-cb31114ab8ed.png)

So the root cause seems because of the `onFailure().recoverWithItem()` method provided:

![98AC0C47-6FF6-446B-ABC3-6FF82A78924B](https://user-images.githubusercontent.com/3773928/148647481-566d5e00-a735-434f-8ea5-79c711de184f.png)

Do you have any idea on this?

**Reply by benkard** [reply]
`null` tends to throw off type inference.  Try casting it to the desired type.

**Comment by Moicen** [comment]
@benkard By casting the `null` in `onItem()` and using `transformToUni()`, seems it works:

![BC61EFB4-B321-4404-9F32-AA21A472F171](https://user-images.githubusercontent.com/3773928/148686510-c4dcfcf0-96cd-4a58-b567-7df5bf913054.png)

I guess this problem is solved, if I have any more problem I’ll start a new thread. Thanks for help! 

In addition here’s all the working combinations if anyone else need it:

```java
// All the working methods

@ServerRequestFilter
public Uni<RestResponse<Void>> check(ContainerRequestContext ctx) throws Exception {
    return Uni.createFrom().nullItem();
}

@ServerRequestFilter
public Uni<RestResponse<Void>> check2(ContainerRequestContext ctx) throws Exception {
    return null;
}

@ServerRequestFilter
public Uni<RestResponse<Void>> check3(ContainerRequestContext ctx) throws Exception {
    return Uni.createFrom().nullItem().onItem().transform(
            _ok -> null
    );
}

@ServerRequestFilter
public Uni<RestResponse<Void>> check4(ContainerRequestContext ctx) throws Exception {
    return Uni.createFrom().nullItem().onItem().transformToUni(
            _ok -> (Uni<RestResponse<Void>>) null
    ).onFailure().recoverWithItem(
            _bad -> null
    );
}

@ServerRequestFilter
public Uni<RestResponse<Void>> check4b(ContainerRequestContext ctx) throws Exception {
    return Uni.createFrom().nullItem().onItem().transformToUni(
            _ok -> (Uni<RestResponse<Void>>) null
    ).onFailure().recoverWithItem(
            _bad -> RestResponse.status(Response.Status.FORBIDDEN)
    );
}

@ServerRequestFilter
public Uni<RestResponse<Void>> check5(ContainerRequestContext ctx) throws Exception {
    return Uni.createFrom().nullItem().onItem().transformToUni(
            _ok -> Uni.createFrom().nullItem().onItem().transform(_ok2 -> null)
    );
}
```

**Comment by Moicen** [comment]
Additional example:

![A278BAC5-ED01-4687-AF9C-54D2ECFCD23C](https://user-images.githubusercontent.com/3773928/148687138-66c6e514-ac9c-4a6c-ad66-0a146d022b03.png)

```java
@ServerRequestFilter
public Uni<RestResponse<Void>> check6(ContainerRequestContext ctx) throws Exception {
    return Uni.createFrom().nullItem().onItem().transformToUni(
            _ok -> Uni.createFrom()
                    .nullItem()
                    .onItem()
                    .transformToUni(_ok2 -> (Uni<RestResponse<Void>>) null)
                    .onFailure().recoverWithItem(
                            err -> null
                    )
    ).onFailure().recoverWithItem(
            err -> null
    );
}
```

**Comment by Moicen** [comment]
After applying what I have learned from this thread, I finally make my filter work.

![986E3213-3BE0-4B43-B82B-B99D4100D522](https://user-images.githubusercontent.com/3773928/148690091-92fc74c4-0417-41d9-ab27-b04aea9660b7.png)

But here is one more note to take:

![A44BCB2C-EF43-4A5F-8960-116652A24844](https://user-images.githubusercontent.com/3773928/148690096-cd493a56-1afe-4356-a671-cdfd0195f711.png)

As the screenshot shown above, I can’t directly use `null` in `onItem().transformToUni()` method, as it will produce mapper error on `null value`.

In addition, I can’t use `Uni.createFrom().nullitem()` too, because it’s implementation is:

![241CAD95-F3AD-4B28-A0DF-589ABD8EEA81](https://user-images.githubusercontent.com/3773928/148690115-a549d18a-9b90-477d-854f-b3fdb5ebade9.png)

![B021B2F5-D855-4E7F-98B7-FA9891AE9C31](https://user-images.githubusercontent.com/3773928/148690124-db6d8720-3733-4325-b3c1-18298a732f2b.png)

And it doesn’t conform to the return type, so I have to cast the type by myself:

```java
return Uni.createFrom().item((RestResponse<Void>) null);
```

At last, in `onFailure().recoverWithItem()` block, I can’t use `Uni.createFrom()` same with `onItem()` method, which is confusing to me, and I need to dig more into detail implementation of Mutiny.

Anyway I have learned a lot and it’s fun!

**Reply by benkard** [reply]
`Uni.createFrom().<RestResponse<Void>>nullItem();`

**Comment by PrakharSrivastav** [comment]
This is really great. Well documented and I learned so much in this single thread than what I did in part week.
